Framework/Core/src/RawDeviceSource.cxx

```diff
@@ -0,0 +1,131 @@
+  }
+}
+
+// FIXME: should I filter out only the output specs which match?
+InjectorFunction o2DMAdaptor(OutputSpec const &spec, uint64_t startTime, uint64_t step) {
+  auto timesliceId = std::make_shared<size_t>(startTime);
+  return [timesliceId, step](FairMQDevice &device, FairMQParts &parts, int index) {
+    for (size_t i = 0; i < parts.Size()/2; ++i) {
+      auto dh = o2::Header::get<DataHeader>(parts.At(i*2)->GetData());
+
+      LOG(DEBUG) << "timesliceId: " << *timesliceId;
+      DataProcessingHeader dph{*timesliceId, 0};
+      o2::Header::Stack headerStack{*dh, dph};
+      broadcastMessage(device, std::move(headerStack), std::move(parts.At(i*2+1)), index);
+      auto oldTimesliceId = *timesliceId;
+      *timesliceId += 1;
+      assert(*timesliceId == oldTimesliceId +1);
+    }
+  };
+}
+
+InjectorFunction incrementalConverter(OutputSpec const &spec, uint64_t startTime, uint64_t step) {
+  auto timesliceId = std::make_shared<size_t>(startTime);
+
+  return [timesliceId, spec, step](FairMQDevice &device, FairMQParts &parts, int index) {
+    // We iterate on all the parts and we send them two by two,
+    // adding the appropriate O2 header.
+    for (size_t i = 0; i < parts.Size(); ++i) {
+      DataHeader dh;
+      dh.dataOrigin = spec.origin;
+      dh.dataDescription = spec.description;
+      dh.subSpecification = spec.subSpec;
+      dh.payloadSize = parts.At(i)->GetSize();
+
+      DataProcessingHeader dph{*timesliceId, 0};
+      *timesliceId += step;
+      //we have to move the incoming data
+      o2::Header::Stack headerStack{dh, dph};
+
+      broadcastMessage(device, std::move(headerStack), std::move(parts.At(i)), index);
+    }
+  };
+}
+
+DataProcessorSpec rawDeviceSource(char const *name,
+                                  std::vector<OutputSpec> const &outputs,
+                                  char const *channelConfig,
+                                  std::function<void(FairMQDevice &device,
+                                                     FairMQParts& inputs,
+                                                     int index)> converter) {
+  DataProcessorSpec spec;
+  spec.name = strdup(name);
+  spec.inputs = {};
+  spec.outputs = outputs;
+  // The Init method will register a new "Out of band" channel and 
+  // attach an OnData to it which is responsible for converting incoming
+  // messages into DPL messages.
+  spec.algorithm = AlgorithmSpec{[converter, channel = spec.name](InitContext &ctx) {
+      auto device = ctx.services().get<RawDeviceService>().device();
+      assert(device);
+
+      // Converter should pump messages
+      auto handler = [device,converter](FairMQParts &inputs, int idx) {
+        converter(*device, inputs, idx);
+        return true;
+      };
+
+      device->OnData(channel, handler);
+      return [](ProcessingContext &ctx) {};
+    }
+  };
+  /// Because of the way FairMQ works, the only hook to create the custom channel
+  /// source is the program options.
```

Comment:
Is that true ? I thought Alexey had fixed the issue(s) with creating FairMQChannel by hand.

Reply:
I can check, it could be there is no other option if you use the simplified API (which I am at the moment). I can switch this to be a `TODO:` comment and explain the situation.